zipline/data/us_equity_loader.py

```diff
@@ -0,0 +1,240 @@
+    ----------
+    window : AdjustedArrayWindow
+       Window of pricing data with prefetched values beyond the current
+       simulation dt.
+    cal_start : int
+       Index in the overall calendar at which the window starts.
+    cal_end : int
+       Index in the overall calendar at which the window
+    """
+
+    def __init__(self, window, cal_start, cal_end):
+        self.window = window
+        self.cal_start = cal_start
+        self.cal_end = cal_end
+        self.current = around(next(window), 3)
+        self.most_recent_ix = self.cal_start
+
+    def get(self, end_ix):
+        """
+        Returns
+        -------
+        out : A np.ndarray of the equity pricing up to end_ix after adjustments
+              and rounding have been applied.
+        """
+        if end_ix < self.most_recent_ix:
+            raise Exception('Can not access data after window has passed.')
+
+        anchor = end_ix - self.cal_start + 1
+        while self.window.anchor < anchor:
+            self.current = around(next(self.window), 3)
+
+        self.most_recent_ix = end_ix
+        return self.current
+
+
+class USEquityHistoryLoader(object):
+    """
+    Loader for sliding history windows of adjusted US Equity Pricing data.
+
+    Parameters
+    ----------
+    daily_reader : DailyBarReader
+        Reader for daily bars.
+    adjustment_reader : SQLiteAdjustmentReader
+        Reader for adjustment data.
+    """
+
+    def __init__(self, daily_reader, adjustment_reader):
+        self._daily_reader = daily_reader
+        self._calendar = daily_reader._calendar
+        self._adjustments_reader = adjustment_reader
+        self._daily_window_blocks = {}
+
+        self._prefetch_length = 40
+
+    def _get_adjustments_in_range(self, assets, days, field):
+        """
+        Get the Float64Multiply objects to pass to an AdjustedArrayWindow.
+
+        For the use of AdjustedAraryWindow in the loader, which looks back
+        from current simulation time back to a window of data the dictionary is
+        structured with:
+        - the key into the dictionary for adjustments is the location of the
+        day from which the window is being viewed.
+        - the start of all multiply objects is always 0 (in each window all
+          adjustments are overlapping)
+        - the end of the multiply object is the location before the calendar
+          location of the adjustment action, making all days before the event
+          adjusted.
+
+        Parameters
+        ----------
+        assets : iterable of Asset
+            The assets for which to get adjustments.
+
+        days : iterable of datetime64-like
+            The days for which adjustment data is needed.
+        field : str
+            OHLCV field for which to get the adjustments.
+
+        Returns
+        -------
+        out : The adjustments as a dict of loc -> Float64Multiply
+        """
+        sids = {int(asset): i for i, asset in enumerate(assets)}
```

Comment:
Might be more semantically correct to use `asset.sid`. This function is only called about once a quarter in a simulation, so any gain would be marginal.